File: src/main/java/com/leijendary/spring/boot/template/model/LocalizedCopy.java
package com.leijendary.spring.boot.template.model;

import static com.leijendary.spring.boot.template.util.RequestContext.getLanguage;
import static java.util.Comparator.comparingInt;
import static java.util.stream.Collectors.toList;
import static javax.persistence.FetchType.EAGER;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

import javax.persistence.CollectionTable;
import javax.persistence.ElementCollection;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.MappedSuperclass;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@Data
@EqualsAndHashCode(callSuper = true)
@MappedSuperclass
public abstract class LocalizedCopy<T extends LocaleCopy> extends AppModel {

    @Id
    private UUID id;

    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @ElementCollection(fetch = EAGER)
    @CollectionTable(joinColumns = @JoinColumn(name = "id"))
    private Set<T> translations = new HashSet<>();

    public T getTranslation() {
        final var language = getLanguage();
        final var sorted = translations
                .stream()
                .sorted(comparingInt(LocaleCopy::getOrdinal))
                .collect(toList());

        return sorted
                .stream()
                .filter(t -> t.getLanguage().equals(language))
                .findFirst()
                .orElse(sorted.iterator().next());
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/model/LocalizedCopy.kt
package com.leijendary.spring.boot.template.core.model

import com.leijendary.spring.boot.template.core.util.RequestContext.language
import java.util.*
import javax.persistence.*

@MappedSuperclass
abstract class LocalizedCopy<T : LocaleCopy> : AppModel() {
    @Id
    var id: UUID? = null

    @ElementCollection
    @CollectionTable(joinColumns = [JoinColumn(name = "id")])
    val translations: Set<T> = HashSet()

    val translation: T
        get() {
            val language: String = language
            val sorted = translations.sortedBy { it!!.ordinal }

            return sorted.first { it!!.language == language } ?: sorted.first()
        }
}